Generate changelog for 3.0.0 (#9554)
tarantool/tarantool: changelogs/3.0.0.md

  the desired bootstrap leader. For example:
  ```lua
    box.cfg{
        bootstrap_strategy = 'config',
        bootstrap_leader = 'localhost:3301',
        replication = {
            'localhost:3301',
            'localhost:3302',
        },
        listen = '3302',
    }
  ```
  (gh-7999).
* You may now control which node new replicas choose as a bootstrap leader
  without touching node config. To do so, set `box.cfg.bootstrap_strategy` to
  `'supervised'`, and the nodes will only bootstrap off the node on which you
  called `box.ctl.make_bootstrap_leader()` last.
  This works on an empty replica set bootstrap as well: start the admin console
  before configuring the nodes. Then configure the nodes:
  ```lua
  box.cfg{
      bootstrap_strategy = 'supervised',
      replication = ...,
      listen = ...,
  }
  ```
  Finally, call `box.ctl.make_bootstrap_leader()` through the admin console
  on the node you want to promote. All the nodes will bootstrap off that node
  (gh-8509).
* A new option `box.cfg.cluster_name` allows assigning a human-readable name to
  the entire cluster. It has to match in all instances and is displayed in
  `box.info.cluster.name` (gh-5029).

> * A new option `box.cfg.replicaset_name` allows assigning a human-readable name
>   to the replicaset. It works the same as `box.cfg.replicaset_uuid`. Its value
>   must be the same across all instances of one replicaset. The replicaset name
>   is displayed in `box.info.replicaset.name` (gh-5029).
> 
> * A new option `box.cfg.instance_name` allows assigning a human-readable name to
>   the instance. It works the same as `box.cfg.instance_uuid`. Its value must be
>   unique in the replicaset. The instance name is displayed in `box.info.name`.
>   Names of other replicas in the same replicaset are visible in
>   `box.info.replication[id].name` (gh-5029).

Reviewer: ```suggestion
* A new option `box.cfg.replicaset_name` allows assigning a human-readable name
  to the replica set. It works the same as `box.cfg.replicaset_uuid`. Its value
  must be the same across all instances of one replica set. The replica set name
  is displayed in `box.info.replicaset.name` (gh-5029).

* A new option `box.cfg.instance_name` allows assigning a human-readable name to
  the instance. It works the same as `box.cfg.instance_uuid`. Its value must be
  unique in the replicaset. The instance name is displayed in `box.info.name`.
  Names of other replicas in the same replica set are visible in
  `box.info.replication[id].name` (gh-5029).
```

Author: updated (you missed one 4th line from bottom), force pushed

  It is called when any event in the trigger registry is modified (gh-8664).
* Introduced helpers for encoding and decoding IPROTO packets in Lua (gh-8054).

#### Msgpack

* Improved error reporting for `msgpack.decode`. Now, an error raised by
  `mgpack.decode` has a detailed error message and the offset in the input
  data. If `msgpack.decode` failed to unpack a MsgPack extension, it also
  includes the error cause pointing to the error in the extension data
  (gh-7986).

#### Netbox

* Introduced the new `net.box` module function `from_fd` for creating a new
  connection from a socket file descriptor number (gh-8984).

#### Popen

* Introduced new option `inherit_fds` for `popen.new`. The option takes
  an array of file descriptor numbers that should be left open in the child
  process (gh-8926).

#### Socket

* Introduced new socket functions `socket.socketpair`, `socket.from_fd`, and
  `socket:detach` (gh-8927).

#### Xlog

* Introduced the `xlog.meta()` method for reading a meta block from an xlog
  file.

### SQL

* Introduced the `SHOW CREATE TABLE` statement (gh-8098).
* **[Breaking change]** The `DROP CONSTRAINT` statement has been
  improved (gh-9112).

> * **[Breaking change]** Names in SQL are now case sensitive. To support backward

Reviewer: ```suggestion
* **[Breaking change]** Names in SQL are now case-sensitive. To support backward
```

Author: ditto

  compatibility, a second lookup using a name normalized using the old rules is
  added (gh-4467).
* `SHOW CREATE TABLE` no longer supports the DEFAULT clause (gh-8793).
* `SQL_EXPR` functions can now be set as a default value (gh-8793).
* A literal set as the default value can no longer have a `+` or `-` sign unless
  the literal is numeric (gh-8793).
* **[Breaking change]** SQL now uses the BOX mechanism for default
  values (gh-8793).

### App

> * Added the new `--force-recovery` CLI option (gh-8876).

Reviewer 2: This section contains just one item regarding a CLI option. I would merge it into the CLI section.

Author: ditto

### Box

* Non-system spaces in a snapshot file are now sorted by their IDs.
  As before, all non-system spaces are stored after system spaces (gh-7954).
* **[Breaking change]** Dropped support for `IPROTO_CALL16` from `net.box`
  (i.e., the `call_16` option and the `call_16` method for `net.box`

>   connections) (gh-8636).

Reviewer 2: Maybe place all the breaking changes at top on the corresponding sections?

Author: done

  `box.schema.user.grant`. Now only users with the `'execute'` privilege
  granted on `sql` or `universe` can execute SQL expressions with the
  `IPROTO_EXECUTE` or `IPROTO_PREPARE` requests. To revert to the old behavior
  (no SQL access checks), use the `sql_priv` compat option (gh-8803).
* Added the `leader_name` field to `box.info.election` (gh-8931).
* Now it is possible to specify the sort order for each part of an index
  (gh-5529).
* Added support for sending tuple formats in IPROTO responses. Added a
  `box_tuple_extension` backward compatibility option to disable sending
  tuple formats in responses to IPROTO call and eval requests (gh-8146).
* Added support for sending tuple formats in IPROTO call and eval request
  arguments. `box_tuple_extension` backward compatibility option can be used to
  disable receiving tuple formats in IPROTO call and eval request arguments
  (gh-8633).
* Added a new `is_sync` parameter to `box.atomic()`. To make the transaction
synchronous, set the `is_sync` option to `true`. Setting `is_sync = false` is
prohibited. If any value other than true/nil is set, for example
`is_sync = "some string"`, then an error will be thrown (gh-8650).

> ### Cli

Reviewer 2: Nit: I would change it to CLI.

Author: ditto

* **[Behavior change]** Disabled starting the Lua REPL by default when running
  Tarantool. Now, Tarantool yields the message that shows the command usage.
  To run the Lua REPL, just set the `-i` flag. To pass a Lua script contents via
  `stdin`, use dash (`-`) as the script name. For more information see a help
  message by running `tarantool -h` (gh-8613).
* Introduced new command-line options `--name` and `--config` (gh-8613).

### Compat

* The following compatibility options' defaults were switched to new behavior:
    * `yaml_pretty_multiline`
    * `sql_seq_scan_default`
    * `json_escape_forward_slash`
    * `fiber_channel_close_mode`
    * `fiber_slice_default`
    * `box_cfg_replication_sync_timeout`
    * `c_func_iproto_multireturn`

  More information on the new behavior can be found on the [Module compat](https://www.tarantool.io/en/doc/latest/reference/reference_lua/compat/) page.

### Console

> * Now multiline commands can be used by setting continuation marker

Reviewer: ```suggestion
* Now multiline commands can be used by setting the continuation marker
```

Author: ditto

  using the `\set continuation` command.

### Fiber

* Exported `fiber_set_name_n`, `fiber_name`, `fiber_id`, `fiber_csw` and
  `fiber_find` into the public C API and usable via FFI as well.
* Make `fiber_set_joinable`, `fiber_set_ctx` and `fiber_get_ctx`
  treat the NULL argument as the current fiber.

### Space

* Introduces space type: a new space definition field.
* Introduces the fully temporary space type. It is the same as data-temporary
  but also has temporary metadata. Temporary spaces can now be created in
  read_only mode, they disappear after server restart and don't exist on
  replicas (gh-8323).

### Test

#### Fuzz

> * Added a grammar-based SQL fuzzer (gh-4826).

Reviewer 2: How about moving it down, after 'build'?

Author: done

### Uri

> * Introduced the `login` and `password` fields in `uri.parse()` (gh-9435).

Reviewer 2: Uri -> URI. However, I would move it to Lua -- here we have just one item.

Author: done, moved to lua

### Build

* Tarantool does not depend on libgomp anymore (gh-7689).
* Added building static `deb` and `rpm` packages.
* Dropped building distro-specific `deb` and `rpm` Tarantool packages.
* Dropped building Tarantool packages with GC32. Now only GC64 packages
  are shipped.
* Exported all LuaJIT public functions from Tarantool binary (gh-3680).

## Bugs fixed

### Core

* **[Breaking change]** The key `box.space._schema['cluster']` is renamed to
  `'replicaset_uuid'`. That is not expected to be breaking because `_schema` is
  an internal system space, but the key was visible in public and documented
  (gh-5029).
* **[Breaking change]** The table `box.info.cluster` is renamed to

>   `box.info.replicaset`. The behaviour can be reverted using the `compat` option

Reviewer: ```suggestion
  `box.info.replicaset`. The behavior can be reverted using the `compat` option
```

Author: ditto

  `box_info_cluster_meaning`

>   (https://tarantool.io/compat/box_info_cluster_meaning) (gh-5029).

Reviewer 2: The page is not created yet: https://github.com/tarantool/doc/issues/3465.

Author: We just started rolling docs. It will appears soon. @andreyaksenov could you double check the address will be valid?

Reviewer: The `https://tarantool.io/compat/box_info_cluster_meaning` page doesn't exist in docs. It is better to add a link to the root `Module compat` topic: https://www.tarantool.io/en/doc/latest/reference/reference_lua/compat/

Reviewer: We can update this as follows:

```
The behaviour can be reverted using the `compat` option
  `box_info_cluster_meaning`
  (https://tarantool.io/compat/box_info_cluster_meaning) (gh-5029).
  
->

The behaviour can be reverted using the [compat](https://www.tarantool.io/en/doc/latest/reference/reference_lua/compat/) option
  `box_info_cluster_meaning` (gh-5029).
```

Author: agreed to remove direct link.

  iterator argument if a transaction was aborted by a fiber yield or by a
  timeout (gh-9340).
* Fixed a bug when `on_rollback` triggers were not invoked during a rollback
  to a savepoint (gh-7810).
* Fixed a bug when the assertion in `box_wait_limbo_acked` would fail. The
  assertion is that the lsn of the last entry in limbo is always positive after
  `wal_sync`. Before the patch, if the `replication_synchro_quorum` was set too
  high on the replica, it would never be reached. After the timeout was
  triggered, the user received a `TimedOut` error. If the quorum was greater
  than or equal to the number of instances in the replica set, the program
  immediately stopped with a segmentation fault (gh-9235).
* Fixed a bug when the assertion in `box_promote_qsync` would fail in the
  debug build mode. The assertion is that at the moment when `box_promote_qsync`
  is called, no other promote is being executed. It turned out that this
  assertion is basically incorrect. In the release build mode, this incorrect
  assumption could potentially lead to writing 2 PROMOTE entries in the same
  term (gh-9263).

### Replication

> * Fixed a bug that allowed to change replicaset UUID via `_schema` space.

Reviewer: ```suggestion
* Fixed a bug that allowed to change a replica set UUID via the `_schema` space.
```

Author: ditto

* Fixed a bug when new instances could try to register via an anon instance
  which previously failed to apply `box.cfg{replication_anon = false}`.
* Fixed a bug when `box.ctl.demote()` with `box.cfg{election_mode = 'off'}`
  and an owned synchro queue could simply not do anything (gh-6860).
* Fixed anonymous replicas not receiving the synchronous transaction queue state
  during join (gh-9401).
* Now transactions are not rolled back if the transaction fiber is

>   cancelled when waiting for quorum from replicas (gh-9480).

Reviewer: ```suggestion
  cancelled when waiting for a quorum from replicas (gh-9480).
```

Author: ditto

  Now the error is thrown when more than 6 alignment settings are pushed on the
  internal stack.
* Fixed incorrect fold rule for `x - (-0)` on trace (for `x = -0` the result
  should be `0`).
* Fixed output for `IR_CONV` in `jit.dump()`.
* Fixed `math.min()`/`math.max()` inconsistencies for x86/x86_64 architectures
  when called with a NaN argument or `-0` and `0`.
* Fixed `math.ceil(x)` result sign for -1 < x < 0.5.
* Errors from gc finalizers are now printed instead of being rethrown.
* Fixed `lua_concat()`.
* Fixed possible storing of NaN keys to table on trace.
* Fixed ABC FOLD optimization with constants.
* Marked `CONV` as non-weak, to prevent invalid control flow path choice.
* Fixed CSE of a `REF_BASE` operand across `IR_RETF`.
* Fixed the fold rule for `BUFHDR APPEND`.
* Fixed HREFK, ALOAD, HLOAD, forwarding vs. `table.clear()`.
* Fixed snapshot PC when linking to `BC_JLOOP` that was a `BC_RET*`.
* Fixed dangling references to CType.
* Ensured returned string is alive in `ffi.typeinfo()`.
* Fixed the missing initialization of the internal structure, leading to a
  crash when recording a trace with an allocation of cdata.

### Lua

> #### Fio

Reviewer 2: I would drop the sub-section (fio), because it has just one item.

Author: done

  transaction to avoid database inconsistency on failed operations (gh-4348).
* **[Breaking change]** `box.schema.user.grant()` now raises an error on
  an attempt to grant the `execute` privilege on a space.  Historically,
  this action was allowed although it had no effect. It's still possible
  to revert to the old behavior with the new compatibility option
  `box_space_execute_priv` (gh-9277).
* Added a check that disables setting `box.cfg` and `log.cfg` options to
  infinite numbers (NaN, Inf). Setting a `box.cfg` or `log.cfg` option to
  an infinite number could result in a crash or invalid behavior (gh-4962).
* Fixed a bug when local references to a space object got out of sync with the
  space on the space drop rollback (gh-9120).
* Fixed a very rare bug when a function deletion could cause an inconsistent
  state of the database (gh-9426).
* Now `box.cfg()` correctly interprets the `TT_LISTEN` and `TT_REPLICATION`
  environment variables with query parameters (gh-9539).

### Http

* Fixed a crash on garbage collection of httpc objects (gh-9283).
* Fixed a crash due to a race in GC finalizers (gh-9346).

> * Fixed behaviour of GC-finalizer of HTTP client Lua object when chunked HTTP

Reviewer: ```suggestion
* Fixed behavior of GC-finalizer of HTTP client Lua object when chunked HTTP
```

Author: ditto
